# Pandas Tutorial

**Tutorial by Ellianna Abrahams**

In addition to being adorable, bamboo-loving bears, **pandas** is a Python library that makes it easy to store and work with large datasets. This notebook will teach you the basics of using pandas.

![](https://cache.desktopnexus.com/thumbseg/1321/1321509-bigthumbnail.jpg)

Over the years many people have been building complex functions in python to allow for analyzing data. These functions can be built into libraries or modules, and there are several that are used across many disciplines. For last week's assignment you worked with numpy, a library containing complex mathematical functions. Today we'll be working with pandas in this tutorial, and matplotlib in the next tutorial. Let's import those libraries here and save them in variables so that we can easily interact with them later.

### Imports

In future notebooks, the following cell will usually be at the top of any code. Generally organizing imports at the top of a notebook makes it easier to remember which libraries you've already imported and follows the guidelines of good coding practice.

In [7]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

Show what is stored inside the variable `np`.

In [8]:
# Which function above would you use to show you what is inside np? Give it a try here!

## The Basics - Create Data

Let's take a look at the 5 most popular baby names in the U.S. in 2018 by. Data was taken from www.ssa.gov.

In [9]:
# We can write words into python code by surrounding a word or number with quotation marks.
# This is called a string, and we have stored our five strings in a list.
names = ['Liam', 'Emma', 'Noah', 'Olivia', 'Ava']

# Whole numbers are called integers.
birth_count = [19837, 18688, 18267, 17921, 14924]

# Decimals or fractions are called floats.
percentage_of_births = [0.0102, 0.0101, 0.0097, 0.0095, 0.0081]

In [10]:
print (type(names))

<class 'list'>


Let's store this data in a pandas spreadsheet, called a `DataFrame`. 
We can't create a dataframe directly from our lists, so we're going to store them in a dictionary, or `dict`.

In [11]:
data = {'names': names, 'count': birth_count, 'percentage': percentage_of_births}

In [12]:
print (type(data))

<class 'dict'>


Now we can make a dataframe! Notice how we access the `DataFrame` function in the pandas library by using the variable `pd` that we defined above.

In [13]:
df = pd.DataFrame(data)

In [44]:
df.head(5)

names  count  percentage
0    Liam  19837      0.0102
1    Emma  18688      0.0101
2    Noah  18267      0.0097
3  Olivia  17921      0.0095
4     Ava  14924      0.0081

We can save our dataframe to a csv file so that we can share it with our colleagues. If you want to save it somewhere specific on your computer, you can specify the path inside the string.

In [15]:
df.to_csv('save_the_pandas.csv')

## The Basics - Get Data

Let's say that your research group already has collected data that they've saved in a csv file. You can open csv files from pandas too using the pandas `read_csv()` function.

In [16]:
# This csv file was created by Hadley Wickham
# and is available at https://github.com/hadley/data-baby-names
names_df = pd.read_csv('baby-names.csv')

There are a few ways you can explore what is in your csv file. For example, you can see what columns are included, you can look at the first few lines, or you can see how many rows, or items, there are in the dataframe.

In [17]:
names_df.columns

# Note that the data for this csv file is taken from the US Social Security website. 
# The SSO has historically assigned a binary gender at birth and is therefore biased in its collection.
# It is always important to keep in mind bias that might be included in data collection!

Index(['year', 'name', 'percent', 'gender'], dtype='object')

In [45]:
names_df.head(10)

# Try chaning the .head call. Can you create a different dataframe as a subset of a
# portion of just the original one?

year     name   percent gender
0  1880     John  0.081541    boy
1  1880  William  0.080511    boy
2  1880    James  0.050057    boy
3  1880  Charles  0.045167    boy
4  1880   George  0.043292    boy
5  1880    Frank  0.027380    boy
6  1880   Joseph  0.022229    boy
7  1880   Thomas  0.021401    boy
8  1880    Henry  0.020641    boy
9  1880   Robert  0.020404    boy

In [19]:
len(names_df)

258000

## Data manipulation
One of the most helpful aspects of pandas is how it can be used to quickly perform operations on large data sets.

In [59]:
# Given a simple dataframe of two columns of values, we can find the sum or
# perform other calculations

column1 = np.random.normal(size=10)
column2 = np.random.normal(size=10)
test_data = {'x': column1, 'y': column2}
dataframe = pd.DataFrame(test_data)
dataframe.head(5)

x         y
0 -0.553867  0.827200
1  0.086062 -0.081750
2  0.526422  0.529994
3 -0.340166  0.636756
4 -2.222968  1.621805

In [60]:
dataframe['sums'] = dataframe['x'] + dataframe['y']
dataframe['product'] = dataframe['x'] * dataframe['y']
dataframe.head(10)

x         y      sums   product
0 -0.553867  0.827200  0.273332 -0.458159
1  0.086062 -0.081750  0.004312 -0.007036
2  0.526422  0.529994  1.056416  0.279001
3 -0.340166  0.636756  0.296589 -0.216603
4 -2.222968  1.621805 -0.601163 -3.605221
5  0.400099 -2.278849 -1.878750 -0.911766
6  1.711945 -0.331535  1.380409 -0.567570
7  2.022670 -0.793661  1.229009 -1.605314
8  0.362749 -0.047591  0.315158 -0.017263
9 -1.181974 -0.591546 -1.773520  0.699192

From exploring the data, we can see that this dataset includes information about 258,000 popular names according to the population percentage given the name each year sorted by gender. 

## Analyzing Data
We can ask questions about this dataset and look for answers using pandas. Let's see how many years are included in the dataset.

In [48]:
# First we isolate the column including information about each year
years = names_df['year']

# Selecting a series like this returns a pandas object called a series
print (type(years))

<class 'pandas.core.series.Series'>


In [50]:
# Now we find the minimum value in that column, using the .min() command
years.min()
print (years.min())

# You can also cast the pandas series object as a list
print (min(list(years)))

1880
1880


Maybe you are interested in knowing how popular the most popular name was by percentage of population given that name, or how popular the least popular name was. What commands above could you use to find the answer to this?

That's a pretty wide range within popularity! But remember that the population of the US has grown since 1880 as well. So maybe a more accurate metric would be to find the average percentage popularity of a name.

In [23]:
names_df['percent'].mean()

0.0008944713333333334

In [24]:
names_df['percent'].median()

0.000164

In [25]:
names_df['percent'].mode()

0    0.000051
dtype: float64

We'll dive more into the difference between mean, median and mode later in the summer. For the final part of this tutorial, we'll look at dividing the data into subsets. Maybe your study is only interested in exploring the most popular names given to baby girls in the 1960s.

In [26]:
subset = names_df.loc[(years >= 1960) & (years < 1970) & (names_df['gender'] == 'girl')]

How do you view the first 12 lines of this subset?

Let's sort our subset by percentage popularity.

In [27]:
subset.sort_values(by='percent', ascending=False)

year       name   percent gender
214000  1965       Lisa  0.032982   girl
215000  1966       Lisa  0.032416   girl
216000  1967       Lisa  0.030539   girl
217000  1968       Lisa  0.028974   girl
212000  1963       Lisa  0.028189   girl
...      ...        ...       ...    ...
209994  1960      Keely  0.000056   girl
209993  1960      Caron  0.000056   girl
209992  1960   Carlotta  0.000056   girl
209991  1960        Bev  0.000056   girl
209999  1960  Anastasia  0.000055   girl

[10000 rows x 4 columns]

## Assignment

Can you make a histogram of the percentage popularity of the name Lisa during the 1960s?

In [42]:
# HINT: It's very useful to break coding down into written steps and then complete each step one by one.

# Here is a sample of the first step that I might write
# Step 1: Create a subset of the name Lisa in the 1960s.
# Step 2: ...
# ...

Can you make a histogram of the percentage popularity of the name Mary during the 1960s?

Now can you plot them on the same histogram?

Please save the csv file of the subset that you created. Scroll to the bottom of [this documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html) to find out how. Once you've completed Ryan's matplotlib tutorial, come back to this one and save a .png of your final histogram too!

## Challenge Assignment

Some names remained popular throughout a decade and some names were only popular for a year or two. 183 of the most popular names in the 1980s were only popular for a year. What were the first five most popular one hit wonders of the 1980s?

We did not go over all the commands that you need to answer this question. Never fear! The internet can be your trusty guide. Head over to https://stackoverflow.com/ and use the pandas documentation to answer this question. Bonus pandas points if you can answer the question in one line of code!

What is the distribution of the count of one hit wonder names over time during the 1980s? Please save your histogram as a .png file.

Now merge this 1980s dataframe with the overall dataset. Can you figure out a way to only keep the names that are popular in and beyond the 1980s in this merge? Write a l